src/util/netif.cpp

```diff
@@ -0,0 +1,275 @@
+                Assume(sizeof(in6_addr) == RTA_PAYLOAD(rta_gateway));
+                in6_addr gw;
+                std::memcpy(&gw, RTA_DATA(rta_gateway), sizeof(gw));
+                return CNetAddr(gw, scope_id);
+            }
+        }
+    }
+
+    return std::nullopt;
+}
+
+#elif defined(WIN32)
+
+#include <iphlpapi.h>
+
+std::optional<CNetAddr> QueryDefaultGateway(Network network)
+{
+    NET_LUID interface_luid = {};
+    SOCKADDR_INET destination_address = {};
+    MIB_IPFORWARD_ROW2 best_route = {};
+    SOCKADDR_INET best_source_address = {};
+    DWORD best_if_idx = 0;
+    DWORD status = 0;
+
+    // Pass empty destination address of the requested type (:: or 0.0.0.0) to get interface of default route.
+    Assume(network == NET_IPV4 || network == NET_IPV6);
+    if (network == NET_IPV4) {
+        destination_address.si_family = AF_INET;
+    } else if(network == NET_IPV6) {
+        destination_address.si_family = AF_INET6;
+    } else {
+        return std::nullopt;
+    }
+
+    status = GetBestInterfaceEx((sockaddr*)&destination_address, &best_if_idx);
+    if (status != NO_ERROR) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Error, "Could not get best interface for default route: %s\n", SysErrorString(status));
+        return std::nullopt;
+    }
+
+    // Get best route to default gateway.
+    // Leave interface_luid at all-zeros to use interface index instead.
+    status = GetBestRoute2(&interface_luid, best_if_idx, nullptr, &destination_address, 0, &best_route, &best_source_address);
+    if (status != NO_ERROR) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Error, "Could not get best route for default route for interface index %d: %s\n",
+                best_if_idx, SysErrorString(status));
+        return std::nullopt;
+    }
+
+    if (network == NET_IPV4) {
+        Assume(best_route.NextHop.si_family == AF_INET);
+        return CNetAddr(best_route.NextHop.Ipv4.sin_addr);
+    } else if(network == NET_IPV6) {
+        Assume(best_route.NextHop.si_family == AF_INET6);
+        return CNetAddr(best_route.NextHop.Ipv6.sin6_addr, best_route.InterfaceIndex);
+    }
+    return std::nullopt;
+}
+
+#elif defined(__APPLE__)
+
+#include <net/route.h>
+#include <sys/sysctl.h>
+
+// Ensure correct alignment for sockaddrs.
+#define ROUNDUP(a) \
+    ((a) > 0 ? (1 + (((a) - 1) | (sizeof(uint32_t) - 1))) : sizeof(uint32_t))
+
+static std::optional<CNetAddr> FromSockAddr(const struct sockaddr* addr)
+{
+    // Fill in a CService from the sockaddr, then drop the port part.
+    CService service;
+    if (service.SetSockAddr(addr)) {
+        return (CNetAddr)service;
+    }
+    return std::nullopt;
+}
+
+std::optional<CNetAddr> QueryDefaultGateway(Network network)
+{
+    // MacOS: Get default gateway from route table.
+    // See man page for route(4) for the format.
+    Assume(network == NET_IPV4 || network == NET_IPV6);
+    int family;
+    if (network == NET_IPV4) {
+        family = AF_INET;
```

Comment:
Alternatively you could store `net.route.0.inet[4].flags.gateway` here and use `sysctlbyname` below, avoiding the need to construct `mib[]`. Not sure if that's better though, because it's nice to be able to lookup constants like `NET_RT_FLAGS` in headers.

Reply:
i prefer using constants to string based APIs, if given the choice. This avoids say, typos.

Reply:
Seems we could factor out `AddressFamilyFromNetwork` at least, this is repeated in literally every implementation 😄